### PR TITLE
Improve generic types support

### LINKED ISSUE #6: import cycles changes when we upgrade the lib to go 1.23 and golangci-lint 1.60.3
```
?   	github.com/synthesio/zconfigcheck/cmd/zconfigcheck	[no test files]
?   	github.com/synthesio/zconfigcheck/golangci	[no test files]
--- FAIL: TestAnalyzer (2.20s)
    --- FAIL: TestAnalyzer/dependency_cycles (0.35s)
        analysistest.go:550: cycles/cycles.go:55:26: diagnostic "configured struct contains dependency cycle: testdata/src/cycles.genericF2 -> testdata/src/cycles.F -> testdata/src/cycles.Generic[testdata/src/cycles.F] -> testdata/src/cycles.F" does not match pattern `configured struct contains dependency cycle: testdata/src/cycles.Generic\[testdata/src/cycles.F\] -> testdata/src/cycles.F -> testdata/src/cycles.Generic\[testdata/src/cycles.F\]`
        analysistest.go:614: cycles/cycles.go:43: no diagnostic was reported matching `configured struct contains dependency cycle: testdata/src/cycles.Generic\[testdata/src/cycles.F\] -> testdata/src/cycles.F -> testdata/src/cycles.Generic\[testdata/src/cycles.F\]`
        analysistest.go:614: cycles/cycles.go:55: no diagnostic was reported matching `configured struct contains dependency cycle: testdata/src/cycles.Generic\[testdata/src/cycles.F\] -> testdata/src/cycles.F -> testdata/src/cycles.Generic\[testdata/src/cycles.F\]`
FAIL
FAIL	github.com/synthesio/zconfigcheck	2.212s
FAIL
```

When I made my tests, I had bumped `golang.org/x/tools` to 0.24.0 https://pkg.go.dev/golang.org/x/tools?tab=versions

### PR DESCRIPTION
**Pull request by dsiragusa**
Use the `golang.org/x/exp/typeparams` package to ensure type assignability is correctly evaluated for generic types.

Resolves #6 by ensuring that all generic type instantiations are analyzed.

### CODE REVIEW

File: structs.go [hunk 1]
```diff
@@ -109,6 +112,28 @@ func (c *checker) checkStructs() {
 			issues.Report(c.Pass)
 		}
 	})
+
+	// This is necessary to catch generic type instantiations.
+	// Generic types have a generic TypeSpec which does not have defined type params.
+	// Because of that, the analysis done on those structs is not complete.
+	//
+	//Furthermore, the generic type does not match the concrete type obtained following the instantiation.
```
Comment (ccoVeille) [inline structs.go:120]:
```suggestion
	// Furthermore, the generic type does not match the concrete type obtained following the instantiation.
```

**Review by ccoVeille — COMMENTED** [review]
Small linting issue

**Review by ccoVeille — DISMISSED** [review]

File: structs.go [hunk 2]
```diff
@@ -109,6 +112,28 @@ func (c *checker) checkStructs() {
...
+
+	// This is necessary to catch generic type instantiations.
+	// Generic types have a generic TypeSpec which does not have defined type params.
+	// Because of that, the analysis done on those structs is not complete.
+	//
+	// Furthermore, the generic type does not match the concrete type obtained following the instantiation.
+	// This means that when we find calls to Configure using instantiated generic types as the argument, we do not
+	// have any available analysis information.
+	//
+	// Running the analysis on all generic type instances avoids these issues.
```
Comment (ccoVeille) [inline structs.go:124]: maybe you could add a comment about the fact the code could be replaced by accessing the field directly

https://pkg.go.dev/golang.org/x/exp/typeparams#GetInstances

https://cs.opensource.google/go/x/exp/+/e7e105de:typeparams/typeparams_go118.go;l=132

Comment (dsiragusa) [reply structs.go:124]: Replaced the call to `GetInstances` with a direct access to `TypesInfo.Instances`

**Review by ccoVeille — DISMISSED** [review]

**Review by ccoVeille — APPROVED** [review]

**Comment by ccoVeille** [comment]
this is a test

**Comment by ccoVeille** [comment]
this is a test

**Comment by ccoVeille** [comment]
this is a test

**Pull request merged**